lib/go-tc/deliveryservices.go

```diff
@@ -192,18 +194,416 @@ type DeliveryServiceFieldsV31 struct {
 // DeliveryServiceV40 is a Delivery Service as it appears in version 4.0 of the
 // Traffic Ops API.
 type DeliveryServiceV40 struct {
-	DeliveryServiceFieldsV31
-	DeliveryServiceFieldsV30
-	DeliveryServiceFieldsV15
-	DeliveryServiceFieldsV14
-	DeliveryServiceFieldsV13
-	DeliveryServiceNullableFieldsV11
+	// Active dictates whether the Delivery Service is routed by Traffic Router.
+	Active bool `json:"active" db:"active"`
```

Comment:
It can't, and it also can't tell the difference between `null` and a missing value - the way Go's JSON parser works by default is highly annoying, and the only real way around it is to write custom marshalling and unmarshalling methods.

I'm going to take those changes out of the PR, they shouldn't have been there in the first place. It just bothers me that assignment and reading always take extra steps for data you know has to exist, is all.

Reply:
Not all pointer-based reads have to be preceded by a nil-check. If it's a required field, generally just the request validation needs to do the nil-check, and the rest of the code can reasonably assume that the field is non-nil.

Reply:
in a handler in TO you can, but then if it's a PUT or POST you need to check again after reading from the database. And utility functions that don't know the contexts in which they might be used need to check, and client methods need to check before access, and code using the clients needs to check before access on the responses, and it needs to get checked ad nauseum in testing code. It's a giant pain that Go won't let you just say "if it's missing this property, throw an error". That would solve so many problems.

Reply:
> It's a giant pain that Go won't let you just say "if it's missing this property, throw an error". That would solve so many problems.

I agree. In fact, I wrote a library to do exactly that: https://github.com/rob05c/apiver - optional things are pointers, non-pointers that are missing throw decode errors. It was rejected on the list.

Reply:
Do you have a link to the list conversation?

Reply:
I'd be more open to having a custom `UnmarshalJSON` method that basically checks the raw JSON as  `map[string]interface{}` and ensures that the required fields are present. After that validation, it could just call `json.Unmarshal` like normal.

Reply:
That's sort of what I was thinking. Either `map[string]interface{}` or you could make "shadow structs" used internally in TO that mirror the API structs, except every property is `json.RawMessage`. Idk which would be easier/smaller.

Reply:
Ideally we'd only have 1 and only 1 struct per resource. Once we have more than 1 struct (like we do today w/ most nullable vs non-nullable structs), they _always_ get out of sync.